dali/kernels/reduce/reduce_all_gpu_test.cu

```diff
@@ -0,0 +1,165 @@
+}
+
+TEST(ReduceGPU, ReduceAllBatchedKernel) {
+  std::mt19937_64 rng(1234);
+  std::uniform_real_distribution<float> dist(0, 1);
+  std::uniform_int_distribution<int> size_dist(10000, 1000000);
+
+  int64_t n_in = 0;
+  int samples = 35;
+  std::vector<int64_t> sizes(samples);
+  for (auto &size : sizes) {
+    size = size_dist(rng);
+    n_in += size;
+  }
+
+  dim3 block(32, 32);
+  int n_out_per_sample = 32;
+  int n_out0 = samples * n_out_per_sample;
+  int n_out = n_out0 + samples;
+  dim3 grid(n_out_per_sample, samples);
+  auto in_data = memory::alloc_unique<float>(AllocType::GPU, n_in);
+  auto out_data = memory::alloc_unique<float>(AllocType::GPU, n_out);
+  std::vector<float> in_cpu(n_in), out_cpu(n_out);
+  for (auto &x : in_cpu)
+    x = dist(rng);
+
+  auto gpu_dev_ptrs = memory::alloc_unique<const float*>(AllocType::GPU, samples);
+  auto gpu_sizes = memory::alloc_unique<int64_t>(AllocType::GPU, samples);
+  vector<const float *> host_ptrs(samples);
+  vector<const float *> cpu_dev_ptrs(samples);
+  int64_t offset = 0;
+  for (int i = 0; i < samples; i++) {
+    host_ptrs[i] = in_cpu.data() + offset;
+    cpu_dev_ptrs[i] = in_data.get() + offset;
+    offset += sizes[i];
+  }
+
+  // data
+  cudaMemcpy(in_data.get(), in_cpu.data(), n_in * sizeof(*in_data), cudaMemcpyHostToDevice);
+  // pointers to sample data
+  cudaMemcpy(gpu_dev_ptrs.get(), cpu_dev_ptrs.data(), samples * sizeof(*gpu_dev_ptrs),
+             cudaMemcpyHostToDevice);
+  // sample sizes
+  cudaMemcpy(gpu_sizes.get(), sizes.data(), samples * sizeof(*gpu_sizes), cudaMemcpyHostToDevice);
+
+  // warm-up
+  ReduceAllBatchedKernel<<<1, block>>>(out_data.get(), gpu_dev_ptrs.get(), gpu_sizes.get());
+  cudaDeviceSynchronize();
+  auto start = CUDAEvent::CreateWithFlags(0);
+  auto end =   CUDAEvent::CreateWithFlags(0);
+  cudaEventRecord(start);
+  ReduceAllBatchedKernel<<<grid, block>>>(out_data.get() + samples,
+                                          gpu_dev_ptrs.get(), gpu_sizes.get());
+
+  dim3 grid2(1, samples);
+  ReduceAllBlockwiseKernel<<<grid2, block>>>(out_data.get(),
+                                             out_data.get() + samples, n_out_per_sample);
+  cudaEventRecord(end);
+  cudaMemcpy(out_cpu.data(), out_data.get(), n_out * sizeof(*out_data), cudaMemcpyDeviceToHost);
+  cudaDeviceSynchronize();
+  float t = 0;
+  cudaEventElapsedTime(&t, start, end);
+
+  offset = 0;
+  for (int i = 0; i < samples; i++) {
+    double ref_sum = RefSum(make_cspan(host_ptrs[i], sizes[i]));
+    double out_sum = out_cpu[i];
+    auto partial_sums = make_cspan(&out_cpu[samples + i * n_out_per_sample], n_out_per_sample);
+    double out_partial = RefSum(partial_sums);
+    EXPECT_NEAR(out_sum, ref_sum, ref_sum * 1e-7 + 1e-7);
+    EXPECT_NEAR(out_partial, ref_sum, ref_sum * 1e-7 + 1e-7);
+    offset += sizes[i];
+  }
+
+  t /= 1000;
+  std::cout << n_in * sizeof(*in_data) / t * 1e-9 << " GB/s" << std::endl;
+}
+
```

Comment:
I think it'd be good to have other reduction here in the tests as an example. For instance, product.

Reply:
Added min, max.